collections/src/main/scala/strawman/collection/ArrayOps.scala

```diff
@@ -79,24 +536,138 @@ class ArrayOps[A](val xs: Array[A]) extends AnyVal
     b.result()
   }
 
+  /** Builds a new array by applying a partial function to all elements of this array
+    * on which the function is defined.
+    *
+    *  @param pf     the partial function which filters and maps the array.
+    *  @tparam B     the element type of the returned array.
+    *  @return       a new array resulting from applying the given partial function
+    *                `pf` to each element on which it is defined and collecting the results.
+    *                The order of the elements is preserved.
+    */
+  def collect[B : ClassTag](f: PartialFunction[A, B]): Array[B] = {
+    var i = 0
+    var matched = true
+    def d(x: A): B = {
+      matched = false
+      null.asInstanceOf[B]
+    }
+    val b = ArrayBuilder.make[B]()
+    while(i < xs.length) {
+      matched = true
+      val v = f.applyOrElse(xs(i), d)
+      if(matched) b += v
+      i += 1
+    }
+    b.result()
+  }
+
   @`inline` final def ++[B >: A : ClassTag](xs: Iterable[B]): Array[B] = appendedAll(xs)
 
-  def zip[B: ClassTag](that: Iterable[B]): Array[(A, B)] = fromTaggedIterable(new View.Zip(this, that))
+  /** Returns an array formed from this array and another iterable collection
+    * by combining corresponding elements in pairs.
+    * If one of the two collections is longer than the other, its remaining elements are ignored.
+    *
+    *  @param   that  The iterable providing the second half of each result pair
+    *  @tparam  B     the type of the second half of the returned pairs
+    *  @return        a new array containing pairs consisting of corresponding elements of this array and `that`.
+    *                 The length of the returned array is the minimum of the lengths of this array and `that`.
+    */
+  def zip[B](that: Iterable[B]): Array[(A, B)] = {
+    val b = new ArrayBuilder.ofRef[(A, B)]()
+    val k = that.knownSize
+    b.sizeHint(if(k >= 0) min(k, xs.length) else xs.length)
+    var i = 0
+    val it = that.iterator()
+    while(i < xs.length && it.hasNext) {
+      b += ((xs(i), it.next()))
+      i += 1
+    }
+    b.result()
+  }
 
-  def zipWithIndex(implicit ct: ClassTag[(A, Int)]): Array[(A, Int)] = fromTaggedIterable(new View.ZipWithIndex(this))
+  /** Zips this array with its indices.
+    *
+    *  @return   A new array containing pairs consisting of all elements of this array paired with their index.
+    *            Indices start at `0`.
+    */
+  def zipWithIndex: Array[(A, Int)] = {
+    val b = new Array[(A, Int)](xs.length)
+    var i = 0
+    while(i < xs.length) {
+      b(i) = ((xs(i), i))
+      i += 1
+    }
+    b
+  }
+
+  /** A copy of this array with an element appended. */
+  def appended[B >: A : ClassTag](x: B): Array[B] = {
+    val dest = new Array[B](xs.length + 1)
+    Array.copy(xs, 0, dest, 0, xs.length)
```

Comment:
@mkeskells suggests using `Arrays.copyOf(xs, xs.length+1, classTag[B])` (https://github.com/scala/scala/pull/6105#issuecomment-345505209)

Reply:
Good idea. We can also optimize all `slice`-based operations with `copyOfRange`.

Reply:
Do you think it's possible to optimize for the case where `B` is `A`, could we have an overload that avoids materializing the ClassTag?